exercises/src/test/scala/datastructures/ListSpec.scala

```diff
@@ -0,0 +1,417 @@
+      List.dropWhile(List(1,2,3))(_ < 4) mustEqual Nil
+      List.dropWhile(List('a','b','c'))(_<'z') mustEqual Nil
+    }
+
+    "remove only first elements matching the predicate in a multiple-element List" in {
+      List.dropWhile(List(5,6,1,2,3))(_ > 4) mustEqual List(1,2,3)
+      List.dropWhile(List('a','b','c'))(_=='a') mustEqual List('b','c')
+    }
+
+    "not remove anything in a List with first element not matching the predicate" in {
+      List.dropWhile(List(5,6,1,2,3))(_ < 4) mustEqual List(5,6,1,2,3)
+      List.dropWhile(List('a','b','c'))(_=='b') mustEqual List('a','b','c')
+    }
+  }
+
+  //========================================================================================
+  //#3.6
+  "init" should {
+    "result in Nil for an empty List" in {
+      List.init(List()) mustEqual Nil
+    }
+
+    "remove the only element in a single-element List" in {
+      List.init(List(1)) mustEqual Nil
+      List.init(List('a')) mustEqual Nil
+    }
+
+    "remove the last element in a multiple-element List" in {
+      List.init(List(1,2,3)) mustEqual List(1,2)
+      List.init(List('a','b','c')) mustEqual List('a','b')
+    }
+    //In order to finish the init call the case Cons(_,Nil) => Nil should match. Which means that we have to call init n-1 times where n is the list length
+  }
+
+  //========================================================================================
+  //#3.8
+  "foldRight" should {
+    "result in original list when passed the case class constructors" in {
+      List.foldRight(List(), Nil:List[Int])(Cons(_,_)) mustEqual List()
+      List.foldRight(List(1,2,3), Nil:List[Int])(Cons(_,_)) mustEqual List(1,2,3)
+    }
+  }
+
+  //========================================================================================
+  //#3.9
+  "length" should {
+    "result in 0 for an empty List" in {
+      List.length(List()) mustEqual 0
+    }
+
+    "result in 1 for a single-element List" in {
+      List.length(List(1)) mustEqual 1
+      List.length(List('a')) mustEqual 1
+    }
+
+    "reslut in the number of elements in a multiple-element List" in {
+      List.length(List(1,2,3)) mustEqual 3
+      List.length(List('a','b','c')) mustEqual 3
+    }
+  }
+
+  // //#3.10
+  // "foldLeft" should {
+  //   "result in original list when passed the case class constructors" in {
+  //     List.foldLeft(List(), Nil:List[Int])((b,a) =>Cons(a,b)) mustEqual List()
+  //     List.foldLeft(List(1,2,3), Nil:List[Int])((b,a) =>Cons(a,b)) mustEqual List(3,2,1)
+  //   }
+  // }
+
+ //========================================================================================
+ //#3.11
+  "sum3 (uses foldLeft)" should {
+    "result in zero for an empty List" in {
+      List.sum3(List()) mustEqual 0
+    }
+
+    "return the element value in a single-element List" in {
+      List.sum3(List(1)) mustEqual 1
+      List.sum3(List('a')) mustEqual 'a'
```

Comment:
Hmm, we shouldn't be trying to test char-to-int implicit conversion; I wouldn't try to depend on that 'feature' in production :-)